source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -90,9 +90,9 @@ LocalRateLimiterImpl::~LocalRateLimiterImpl() {
 }
 
 void LocalRateLimiterImpl::onFillTimer() {
+  refill_counter_++;
   onFillTimerHelper(tokens_, token_bucket_);
   onFillTimerDescriptorHelper();
-  refill_counter_++;
```

Comment:
Is this change necessary?

Reply:
Actually, I have changed it due to the logic utilized to check whether the bucket refil should occur. Bellow is a snippet of the code

```c++
void LocalRateLimiterImpl::onFillTimerDescriptorHelper() {
  for (const auto& descriptor : descriptors_) {
    // Descriptors are refilled every Nth timer hit where N is the ratio of the
    // descriptor refill interval over the global refill interval. For example,
    // if the descriptor refill interval is 150ms and the global refill
    // interval is 50ms, this descriptor is refilled every 3rd call.
    if (refill_counter_ % descriptor.multiplier_ == 0) {
      onFillTimerHelper(*descriptor.token_state_, descriptor.token_bucket_);
    }
  }
}
```

Note that the refill will occur whenever the remainder of the division of `refill_counter_` by `descriptor.multiplier_` is zero. However since `refill_counter` is initialized with zero, all descriptor buckets will be refilled the first time the method executes. For example, consider the rate limit configuration below:

```yaml
- match:
    prefix: "/"
  route:
    cluster: app
    priority: "DEFAULT"
    rate_limits:
    - actions:
      - request_headers:
          header_name: x-client-application
          descriptor_key: client_application
          skip_if_absent: true
  typed_per_filter_config:
    envoy.filters.http.local_ratelimit:
      "@type": type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit
      stat_prefix: local_rate_limiter
      token_bucket:
        max_tokens: 10
        tokens_per_fill: 10
        fill_interval: 10s
      filter_enabled:
        runtime_key: local_rate_limit_enabled
        default_value:
          numerator: 100
          denominator: HUNDRED
      filter_enforced:
        runtime_key: local_rate_limit_enforced
        default_value:
          numerator: 100
          denominator: HUNDRED
      descriptors:
      - entries:
        - key: client_application
          value: app1
        token_bucket:
          max_tokens: 2
          tokens_per_fill: 2
          fill_interval: 30s
      - entries:
        - key: client_application
          value: app2
        token_bucket:
          max_tokens: 4
          tokens_per_fill: 4
          fill_interval: 30s
```

In this case, the default bucket will be refilled every 10 seconds and the descriptors bucket every 30s.

The method `LocalRateLimiterImpl::onFillTimer` will be called after 10 seconds, and te default bucket will be refilled as it should. Since `refill_counter_` is zero all descriptor buckets will also be refilled although they should only be after 30 seconds.

Executing `refill_counter_++;` before the method `onFillTimerDescriptorHelper` will ensure that descriptor buckets will be refilled every Nth timer hit.

Reply:
get it. Thanks for your detailed explanation.

Reply:
Can you add some comments in this function if the order matters this much? It will help future readers understand and not move it.

Reply:
For sure.